### PR TITLE
refactor: inject var.xx and res.xx.output attribute before validate

### PR DESCRIPTION
**Pull request by aiwantaozi**
**Problem:**
attributes include ${var.xx} or ${res.xx.output} may failed to validate

**Solution:**
inject attributes before validating with schema

**Related Issue:**
#1931

### CODE REVIEW

File: pkg/apis/resource/helper.go [hunk 1]
```diff
@@ -1,11 +1,187 @@
...
+
+		n := fmt.Sprintf(resourceFormat, ss[0], ss[1])
+		values[n] = v.Value
+	}
+
+	return values, nil
+}
+
+func getShouldInject(
+	bs []byte,
```
Comment (gitlawr) [inline pkg/apis/resource/helper.go:154]: use readable param name

File: pkg/apis/resource/helper.go [hunk 2]
```diff
@@ -1,11 +1,187 @@
...
+			resource.NameIn(resNames...),
+			resource.ProjectID(projectID),
+			resource.EnvironmentID(environmentID)).
+		IDs(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get should inject resource outputs.
+	outputs, err := terraform.GetServiceDependencyOutputs(ctx, client, depIDs, resOutputs)
```
Comment (gitlawr) [inline pkg/apis/resource/helper.go:128]: Calling deployer/terraform pkg in apis/resource seems to be odd. Should we move the function to pkg/resource(and change its name)?

**Review by gitlawr — DISMISSED** [review]
LGTM

File: pkg/apis/resource/helper.go [hunk 3]
```diff
@@ -1,11 +1,188 @@
...
 	"github.com/seal-io/walrus/pkg/dao/types/property"
+	pkgresource "github.com/seal-io/walrus/pkg/resource"
+	pkgvariable "github.com/seal-io/walrus/pkg/variable"
+	"github.com/seal-io/walrus/utils/json"
 )
 
+var (
+	// _variableReg the regexp to match the variable.
+	variableReg = regexp.MustCompile(`\${var\.([a-zA-Z0-9_-]+)}`)
+	// _resourceReg the regexp to match the resource output.
```
Comment (alexcodelf) [inline pkg/apis/resource/helper.go:24]: There are two regexp `\${res\.([^.}]+)\.([^.}]+)` and `\${var\.([a-zA-Z0-9_-]+)}` in this package and `pkg/deployer/terraform` can we unify them? Please update the comment of the variable.

**Review by alexcodelf — APPROVED** [review]
LGTM

**Pull request merged**